csharp/private/tools/assembly-runner/lib.bzl

```diff
@@ -0,0 +1,147 @@
+load(
+    "@d2l_rules_csharp//csharp/private:providers.bzl",
+    "AnyTargetFramework",
+    "CSharpAssembly",
+)
+
+def _write_wrapper_cc_impl(ctx):
+    provider = CSharpAssembly[ctx.attr.target_framework]
+    exe = ctx.attr.exe[provider]
+
+    toolchain = ctx.toolchains["@d2l_rules_csharp//csharp/private:toolchain_type"]
+
+    ctx.actions.expand_template(
+        template = ctx.files._main_cc[0],
+        output = ctx.outputs.out,
+        substitutions = {
+            "{TargetExe}": exe.out.short_path,
+
+            # Strip the leading "../" off the path to dotnet.exe
+            "{DotnetExe}": toolchain.runtime.short_path[3:],
+
+            "{RuntimeConfigJson}": ctx.file.runtimeconfig.short_path,
+        },
+    )
+
+_write_wrapper_cc = rule(
+    _write_wrapper_cc_impl,
+    doc = "Write a .cc file that will wrap a C# exe",
+    attrs = {
+        "exe": attr.label(
+            doc = "The exe to wrap",
+            providers = AnyTargetFramework,
+        ),
+        "target_framework": attr.string(
+            doc = "The target framework to choose at runtime",
+            mandatory = True,
+        ),
+        "out": attr.output(
+            doc = "The .cc file that gets output",
+        ),
+        "runtimeconfig": attr.label(
+            doc = "A *.runtimeconfig.json file",
+            allow_single_file = True,
+            mandatory = True,
+        ),
+        "_main_cc": attr.label(
+            doc = "The .cc file to use as a template",
+            allow_single_file = [".cc"],
+            default = ":main.cc",
+        ),
+    },
+    toolchains = ["@d2l_rules_csharp//csharp/private:toolchain_type"],
+)
+
+def _csharp_exe_impl(ctx):
+    # Collect the C# providers from the exe
+    csharp_providers = []
+    for tf in CSharpAssembly.values():
+        if tf in ctx.attr.exe:
+            csharp_providers += [ctx.attr.exe[tf]]
+
+    # Get the specific provider for the one we want to run
+    provider = CSharpAssembly[ctx.attr.target_framework]
+    exe = ctx.attr.exe[provider]
+
+
+    # TODO: use bat for windows
+    # We need to copy the wrapper because otherwise Bazel responds with this:
+    #  'executable' provided by an executable rule '_csharp_exe' should be created by the same rule.
+    # :(
```

Comment:
This buddy does a cross-platform copy: https://github.com/bazelbuild/bazel-skylib/blob/master/rules/copy_file.bzl

We can't use it directly (because it's a rule, and we want to copy the file within this existing rule specifically).